src/stan-users-guide/parallelization.Rmd

```diff
@@ -1,12 +1,314 @@
+3. ```x_slice``` - The subset of ```x``` (from ```reduce_sum``` / `reduce_sum_static`) for
+which this partial sum is responsible (```x_slice = x[start:end]```)
+4. ```s1, s2, ...``` - Arguments shared in every term  (passed on
+without modification from the ```reduce_sum``` / `reduce_sum_static` call)
+
+The user-provided function ```f``` is expected to compute the partial
+sum with the terms ```start``` through ```end``` of the overall
+sum. The user function is passed the subset ```x[start:end]``` as
+```x_slice```. ```start``` and  ```end``` are passed so that ```f```
+can index any of the tailing ```sM``` arguments as necessary. The
+trailing ```sM``` arguments are passed without modification to every
+call of ```f```.
+
+A ```reduce_sum``` (or `reduce_sum_static`) call:
+
+```
+real sum = reduce_sum(f, x, grainsize, s1, s2, ...);
+```
+
+can be replaced by either:
+
+```
+real sum = f(1, size(x), x, s1, s2, ...);
+```
+
+or the code:
+
+```
+real sum = 0.0;
+for(i in 1:size(x)) {
+  sum += f(i, i, { x[i] }, s1, s2, ...);
+}
+```
+
+### Example: Logistic Regression
+
+Logistic regression is a useful example to clarify both the syntax
+and semantics of reduce summation and how it can be used to speed up a typical
+model.
+
+A basic logistic regression can be coded in Stan as:
+
+```
+data {
+  int N;
+  int y[N];
+  vector[N] x;
+}
+parameters {
+  vector[2] beta;
+}
+model {
+  beta ~ std_normal();
+  y ~ bernoulli_logit(beta[1] + beta[2] * x);
+}
+```
+
+In this model predictions are made about the `N` outputs `y` using the
+covariate `x`. The intercept and slope of the linear equation are to be estimated.
+
+The key point to getting this calculation to use reduce summation, is recognizing that
+the statement:
+
+```
+y ~ bernoulli_logit(beta[1] + beta[2] * x);
+```
+
+can be rewritten (up to a proportionality constant) as:
+```
+for(n in 1:N) {
+  target += bernoulli_logit_lpmf(y[n] | beta[1] + beta[2] * x[n])
+}
+```
+
+Now it is clear that the calculation is the sum of a number of conditionally
+independent Bernoulli log probability statements, which is the condition where
+reduce summation is useful.
+
+To use the reduce summation, a function must be written that can be used to compute
+arbitrary partial sums of the total sum.
+
+Using the interface defined in [Reduce-Sum](#reduce-sum), such a function
+can be written like:
+
+```
+functions {
+  real partial_sum(int start, int end,
+                   int[] y_slice,
+                   vector x,
+                   vector beta) {
+    return bernoulli_logit_lpmf(y_slice | beta[1] + beta[2] * x[start:end]);
+  }
+}
+```
+
+And the likelihood statement in the model can now be written:
+
+```
+target += partial_sum(1, N, y, x, beta); // Sum terms 1 to N of the likelihood
+```
+
+In this example, `y` was chosen to be sliced over because there
+is one term in the summation per value of `y`. Technically `x` would  have
+worked as well. Use whatever conceptually makes the most sense.
```

Comment:
"conceptually makes the most sense" - make the model code easy to understand and maintain?